xdsl/transforms/convert_memref_to_ptr.py

```diff
@@ -228,37 +224,28 @@ def match_and_rewrite(self, op: func.CallOp, rewriter: PatternRewriter, /):
             return
 
         # rewrite arguments
-        insert_point = InsertPoint.before(op)
         new_arguments: list[SSAValue] = []
 
         # insert `memref -> ptr` casts for memref arguments values
         for argument in op.arguments:
             if isinstance(argument.type, memref.MemRefType):
-                rewriter.insert_op(
-                    cast_op := builtin.UnrealizedConversionCastOp.get(
-                        [argument], [ptr.PtrType()]
-                    ),
-                    insert_point,
-                )
-                new_arguments.append(cast_op.results[0])
+                rewriter.insert_op_before_matched_op(cast_op := ptr.ToPtrOp(argument))
+                new_arguments.append(cast_op.res)
+                cast_op.res.name_hint = argument.name_hint
             else:
                 new_arguments.append(argument)
 
-        insert_point = InsertPoint.after(op)
         new_results: list[SSAValue] = []
 
         #  insert `ptr -> memref` casts for return values
         for result in op.results:
-            if isinstance(result.type, memref.MemRefType):
-                rewriter.insert_op(
-                    cast_op := builtin.UnrealizedConversionCastOp.get(
-                        [result],
-                        # TODO: annoying pyright warnings - Sasha, pls help
-                        [result.type],  # pyright: ignore[reportUnknownMemberType,reportUnknownArgumentType]
-                    ),
-                    insert_point,
+            if isa(result.type, memref.MemRefType):
+                rewriter.insert_op_after_matched_op(
+                    cast_op := ptr.FromPtrOp(result, result.type)
                 )
-                new_results.append(cast_op.results[0])
+                new_results.append(cast_op.res)
+                cast_op.res.name_hint = result.name_hint
+                result.replace_by_if(cast_op.res, lambda x: x.operation is not cast_op)
```

Comment:
@alexarice, this is now the only `replace_by_if` added in this PR, and it avoids a bug where the operand to the cast op (result) was rewritten to the result of the cast op

Reply:
I'm sorry, I really don't get what's going on here. `new_results` looks completely unused, and was it just bugged that we didn't have actually use the result of the cast before?

I'd far prefer to collect the new cast ops into a list and then do one big `replace_matched_op` with `(new_call_op, *cast_ops)` and `new_results`

Reply:
Refactored following your suggestion.